### ISSUE TITLE
A ring attention with flash attention kernel implementation

### ISSUE THREAD

**Issue by zhuzilin** [lucidrains/ring-attention-pytorch#4]
Hi! Thank you for your work on implementing the ring attention in pytorch!

I've just tried to implement a `ring_flash_attn_qkvpacked_func` (corresponding to `flash_attn_qkvpacked_func` in flash attention) with the flash attention kernels here: https://github.com/zhuzilin/ring-flash-attention/

Maybe this can help :)

---

Updates:

- `ring_flash_attn_varlen_qkvpacked_func` is also implemented.

**Comment by lucidrains** [comment]
@zhuzilin if you do embark on the pull request, the minimal features would be the ring IPC, able to specify the maximum number of ring passes (as I believe they must have curriculum learned the local attention to a full global, or mixed local and global using variable ring passes throughout the transformer), and finally, if you have the bandwidth, specialize masking logic for striped attention to balance the workload

**Comment by lucidrains** [comment]
<img width="287" alt="Screen Shot 2024-02-21 at 7 11 49 AM" src="https://github.com/lucidrains/ring-attention-pytorch/assets/108653/0120fbcb-504c-4657-b310-7252cd60b1b0">

thank you! 🚀 ❤️

**Comment by lucidrains** [comment]
@zhuzilin actually, after looking into CUDA IPC stuff, your approach may be the best for now

**Comment by zhuzilin** [comment]
> Are you planning on upstreaming the finalized implementation to Tri Dao's official flash attention repository?

I'll draft an issue to the flash attention repo to see if they have interest in merging (or design a better version) in the official repo :)

> after looking into CUDA IPC stuff, your approach may be the best for now
yeah, using nccl based p2p communication would be at least an easier way to implement with acceptable performance.

**Comment by andreaskoepf** [comment]
@zhuzilin awesome work, we‘ll organize a little hack today 19:00 UTC on the cuda-mode discord to hack your impl - just fyi https://x.com/neurosp1ke/status/1760558683136589983

**Comment by lucidrains** [comment]
![8gp0cg](https://github.com/lucidrains/ring-attention-pytorch/assets/108653/a23e9d7a-3764-4e8a-8755-29b96c26eb8e)

**Comment by lucidrains** [comment]
i also wanted to do some LOTR references, but one meme is enough

**Comment by zhuzilin** [comment]
oh... sorry, I took a day off and missed all the notification from github....

**Comment by lucidrains** [comment]
@zhuzilin i think my version is working too now, with a modified forward flash attention kernel to minimize ring passes

thanks for sharing your repo for proof of concept!

**Comment by andreaskoepf** [comment]
@lucidrains thanks a lot for your hard work & very interesting that you used a custom triton kernel! :-)

**Comment by lucidrains** [comment]
@andreaskoepf thanks! seems like there's still issue with backwards, but i'll leave it to someone or some team to fix. yup, i think the forwards requires the key, values to be iterated on the outer loop (to save on extraneous ring passes), so the reduced values needs to be stored and passed back in on the next ring pass. but i could be wrong and there may be a simpler way

**Comment by lucidrains** [comment]
it isn't correct, probably something small with regards to how i'm using the flash attention api

feel free to submit a PR, i likely won't be able to get to this as i'll be running around bay area meeting people next month

**Comment by lucidrains** [comment]
@ericauld ah, good news, the cuda backwards actually yielded the right gradients (full attention, no causal or key padding mask). it is my naive version that is broken

alright, i guess it is safe to remove the wip

**Comment by apaz-cli** [comment]
@lucidrains Knowing the LSE doesn't actually help you compute the backwards for softmax though, correct? The derivative of LSE is softmax, not the other way around. What am I missing, and what is the utility of returning the LSE?

**Comment by andreaskoepf** [comment]
> What am I missing, and what is the utility of returning the LSE?

The returned log sum exp is what allows to apply flash-attenion in a blockwise manner (e.g. without it it wouldn't be possible to use flash-attn to implement ring-attn). See [ring_flash_attn/utils.py#L19-L21](https://github.com/zhuzilin/ring-flash-attention/blob/78959746e8ce88394ded9263b417ec4708f3cc45/ring_flash_attn/utils.py#L19-L21).

**Comment by apaz-cli** [comment]
Ah, alright. That's what I'm missing. Makes sense :)
